crates/ruff/src/args.rs

```diff
@@ -515,101 +554,184 @@ impl From<&LogLevelArgs> for LogLevel {
     }
 }
 
+/// Struct to encapsulate all configuration-related arguments passed via the CLI.
+///
+/// Three kinds of configuration arguments can be provided via the CLI;
+/// each kind is represented by a separate field on this struct.
+#[derive(Default)]
+pub struct ConfigArguments {
+    /// Path to a pyproject.toml or ruff.toml config file (etc.).
+    /// Either 0 or 1 config file paths may be provided on the command line.
+    config_file: Option<PathBuf>,
+    /// Overrides provided via the `--config "KEY=VALUE"` option.
+    /// An arbitrary number of these overrides may be provided on the command line.
+    /// These overrides take precedence over all configuration files,
+    /// even config files that were also specified using `--config`.
+    overrides: Configuration,
+    /// Overrides provided via dedicated flags such as `--line-length` etc.
+    /// These overrides take precedence over all configuration files,
+    /// and also over all overrides specified using any `--config "KEY=VALUE"` flags.
+    per_flag_overrides: ExplicitConfigOverrides,
+}
+
+impl ConfigArguments {
+    pub fn config_file(&self) -> Option<&Path> {
+        self.config_file.as_deref()
+    }
+
+    fn from_cli_arguments(
+        config_options: Vec<SingleConfigArgument>,
+        per_flag_overrides: ExplicitConfigOverrides,
+        isolated: bool,
+    ) -> anyhow::Result<Self> {
+        let mut new = Self {
+            per_flag_overrides,
+            ..Self::default()
+        };
+
+        for option in config_options {
+            match option {
+                SingleConfigArgument::SettingsOverride(overridden_option) => {
+                    let overridden_option = Arc::try_unwrap(overridden_option)
+                        .unwrap_or_else(|option| option.deref().clone());
+                    new.overrides = new.overrides.combine(Configuration::from_options(
+                        overridden_option,
+                        None,
+                        &path_dedot::CWD,
+                    )?);
+                }
+                SingleConfigArgument::FilePath(path) => {
+                    if isolated {
+                        bail!(
+                            "\
+The argument `--config={}` cannot be used with `--isolated`
+
+  tip: You cannot specify a configuration file and also specify `--isolated`,
+       as `--isolated` causes ruff to ignore all configuration files.
+       For more information, try `--help`.
+",
+                            path.display()
+                        );
```

Comment:
I wonder if we should use structured errors for these e.g. introduce an `Error` enum and define the messages there. I'm not sure what the current state of that is for our Ruff CLI.

cc @MichaReiser

Reply:
Micha and I had a longish discussion about this in https://github.com/astral-sh/ruff/pull/9599#discussion_r1469202408. I originally used `clap::error::Error` here, for precisely this reason, but Micha successfully persuaded me that it was a bad idea to use `clap` after the initial argparsing phase :)

I'd quite like to look at introducing more structured errors generally so that it's easier for us to distinguish between errors where the user input from the CLI is the problem and errors that are "our fault". I think there are other places where we use `anyhow` for bubbling up CLI errors that we detect after the initial argparsing; I'd quite like to look at it holistically and defer it to a future issue/PR, if that's laright?

Reply:
Ah yeah I'm thinking you'd just `thiserror` but I am still an error handling noob in Rust. Definitely feel free to defer, especially if you already talked with Micha about it.